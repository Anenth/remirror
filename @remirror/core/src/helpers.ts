import { freeze, keys } from '@remirror/core-helpers';

import { ChangedProperties } from './types';

export interface GetChangedPropertiesParameter<Properties extends object> {
  /**
   * The previous readonly properties object.
   */
  previous: Required<Readonly<Properties>>;

  /**
   * The partial update object that was passed through.
   */
  update: Partial<Properties>;

  /**
   * A method to check whether two value are equal.
   */
  equals?: (valueA: unknown, valueB: unknown) => boolean;
}

function defaultEquals(valueA: unknown, valueB: unknown) {
  return valueA === valueB;
}

/**
 * Get the property changes and the next value from an update.
 */
export function getChangedProperties<Properties extends object>(
  parameter: GetChangedPropertiesParameter<Properties>,
): GetChangedPropertiesReturn<Properties> {
  const { previous, update, equals = defaultEquals } = parameter;
  const next = freeze({ ...previous, update });
  const changes = {} as ChangedProperties<Properties>;
  const propertyKeys = keys(previous);

  for (const key of propertyKeys) {
    const previousValue = previous[key];
    const nextValue = next[key];

    if (equals(previousValue, nextValue)) {
      changes[key] = { changed: false };
      continue;
    }

    changes[key] = { changed: true, previous: previousValue, next: nextValue };
  }

  return { changes: freeze(changes), next };
}

export interface GetChangedPropertiesReturn<Properties extends object> {
  /**
   * The next value of the properties after the update.
   */
  next: Readonly<Required<Properties>>;

  /**
   * An object with all the keys showing what's been changed.
   *
   * @remarks
   *
   * Using this can prevent unnecessary updates. It's possible for new
   * properties to be passed that are identical to the previous, by checking if
   * the object was changed this can be avoided.
   */
  changes: Readonly<ChangedProperties<Properties>>;
}
