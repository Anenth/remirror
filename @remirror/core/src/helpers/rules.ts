import { InputRule } from 'prosemirror-inputrules';
import { Fragment, MarkType, Node as PMNode, Slice } from 'prosemirror-model';
import { Plugin } from 'prosemirror-state';
import { Attrs, GetAttrs, InputRuleCreator } from '../types';
import { findMatches, isFunction } from './base';

/**
 * Creates an input rule based on the provided regex for the provided mark type
 *
 * @param regex - the regex matcher
 * @param type - the mark type
 * @param getAttrs - the attributes or attribute creating function
 *
 * @public
 */
export const markInputRule = (
  regex: RegExp,
  type: MarkType,
  getAttrs?: ((attrs: string[]) => Attrs) | Attrs,
) => {
  return new InputRule(regex, (state, match, start, end) => {
    const attrs = isFunction(getAttrs) ? getAttrs(match) : getAttrs;
    const { tr } = state;
    let markEnd = end;

    if (match[1]) {
      const startSpaces = match[0].search(/\S/);
      const textStart = start + match[0].indexOf(match[1]);
      const textEnd = textStart + match[1].length;
      if (textEnd < end) {
        tr.delete(textEnd, end);
      }
      if (textStart > start) {
        tr.delete(start + startSpaces, textStart);
      }
      markEnd = start + startSpaces + match[1].length;
    }

    tr.addMark(start, markEnd, type.create(attrs));
    tr.removeStoredMark(type); // Do not continue with mark.
    return tr;
  });
};

/**
 * Creates a paste rule based on the provided regex for the provided mark type
 *
 * @param regex - the regex matcher
 * @param type - the mark type
 * @param getAttrs - the attributes or attribute creating function
 *
 * @public
 */
export const markPasteRule = (regexp: RegExp, type: MarkType, getAttrs?: GetAttrs) => {
  const handler = (fragment: Fragment) => {
    const nodes: PMNode[] = [];

    fragment.forEach(child => {
      if (child.isText) {
        const text = child.text || '';
        let pos = 0;

        // ? For some reason including the index param makes this work. I'm not sure why...?
        findMatches(text, regexp).forEach((match, _) => {
          if (match[1]) {
            const start = match.index;
            const end = start + match[0].length;
            const attrs = isFunction(getAttrs) ? getAttrs(match) : getAttrs;

            if (start > 0) {
              nodes.push(child.cut(pos, start));
            }

            nodes.push(child.cut(start, end).mark(type.create(attrs).addToSet(child.marks)));

            pos = end;
          }
        });

        if (text && pos < text.length) {
          nodes.push(child.cut(pos));
        }
      } else {
        nodes.push(child.copy(handler(child.content)));
      }
    });

    return Fragment.fromArray(nodes);
  };

  return new Plugin({
    props: {
      transformPasted: slice => new Slice(handler(slice.content), slice.openStart, slice.openEnd),
    },
  });
};

/**
 * Creates an node input rule based on the provided regex for the provided node type
 *
 * @param regex - the regex matcher
 * @param type - the node type
 * @param getAttrs - the attributes or attribute creating function
 *
 * @public
 */
export const nodeInputRule: InputRuleCreator = (regexp, type, getAttrs) => {
  return new InputRule(regexp, (state, match, start, end) => {
    const attrs = isFunction(getAttrs) ? getAttrs(match) : getAttrs;
    const { tr } = state;

    if (match[0]) {
      tr.replaceWith(start - 1, end, type.create(attrs!));
    }

    return tr;
  });
};

/**
 * Creates an enhanced node input rule based on the provided regex for the provided node type.
 *
 * @remarks
 * This is different from the node input rule in that it checks that the start end end are both
 * valid and replaces from the start position.
 *
 * @param regex - the regex matcher
 * @param type - the node type
 * @param getAttrs - the attributes or attribute creating function
 *
 * @public
 */
export const enhancedNodeInputRule: InputRuleCreator = (regexp, type, getAttrs) => {
  return new InputRule(regexp, (state, match, start, end) => {
    end = start > end ? start : end;
    const attrs = isFunction(getAttrs) ? getAttrs(match) : getAttrs;
    const { tr } = state;

    const str = match[0];

    if (str) {
      tr.replaceWith(start, end, type.create(attrs!));
    }

    return tr;
  });
};
