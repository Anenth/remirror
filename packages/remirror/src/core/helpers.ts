export * from '@remirror/core-helpers';
