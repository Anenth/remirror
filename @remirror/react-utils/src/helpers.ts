import { jsx } from '@emotion/core';
import {
  AnyFunction,
  bool,
  isArray,
  isFunction,
  isObject,
  isString,
  PlainObject,
  uniqueArray,
} from '@remirror/core';
import { Fragment, isValidElement, LegacyRef, PropsWithChildren, ReactElement, ReactNode } from 'react';
import { RemirrorComponentType, RemirrorElement, RemirrorElementType } from './types';

/**
 * Check whether a react node is a built in dom element (i.e. `div`, `span`)
 *
 * @param value - the value to check
 */
export const isReactDOMElement = <GProps extends {} = any>(
  value: unknown,
): value is ReactElement<GProps> & { type: string } => {
  return isObject(value) && isValidElement(value) && isString(value.type);
};

/**
 * Checks whether the element is a react fragment
 *
 * @param value - the value to check
 */
export const isReactFragment = <GProps extends {} = any>(
  value: unknown,
): value is ReactElement<GProps> & { type: typeof Fragment } =>
  isObject(value) && isValidElement(value) && value.type === Fragment;

/**
 * Retrieve the element props for JSX Element
 *
 * @param element
 */
export const getElementProps = (element: JSX.Element): PlainObject & { children: JSX.Element } => {
  return element ? element.props : {};
};

/**
 * Utility for generating a unique class name
 *
 * @param uid
 * @param className
 */
export const uniqueClass = (uid: string, className: string) => `${className}-${uid}`;

/**
 * Utility for properly typechecking static defaultProps for a class component in react.
 *
 * ```ts
 * static defaultProps = asDefaultProps<RemirrorProps>()({
 *   initialContent: EMPTY_OBJECT_NODE,
 * });
 * ```
 */
export const asDefaultProps = <GProps extends {}>() => <GDefaultProps extends Partial<GProps>>(
  props: GDefaultProps,
): GDefaultProps => props;

/**
 * Checks if this element has a type of any RemirrorComponent
 *
 * @param value - the value to check
 */
export const isRemirrorElement = <GOptions extends {} = any>(
  value: unknown,
): value is RemirrorElement<GOptions> => {
  return bool(
    isObject(value) &&
      isValidElement(value) &&
      (value.type as RemirrorComponentType<GOptions>).$$remirrorType,
  );
};

const isRemirrorElementOfType = (type: RemirrorElementType) => <GOptions extends {} = any>(
  value: unknown,
): value is RemirrorElement<GOptions> => isRemirrorElement(value) && value.type.$$remirrorType === type;

/**
 * Checks to see if this is the wrapper we've created around the RemirrorContent.Provider component.
 *
 * This is used to help determine how the Remirror component will be rendered. `getRootProps` is the main reason
 * for this, and I'm not even sure the effort is worth it.
 *
 * @param value - the value to check
 */
export const isRemirrorContextProvider = isRemirrorElementOfType(RemirrorElementType.ContextProvider);

/**
 * Checks if this is a RemirrorExtension type. These are used to configure the extensions that determine
 * the underlying behaviour of the editor.
 *
 * @param value - the value to check
 */
export const isRemirrorExtension = isRemirrorElementOfType(RemirrorElementType.Extension);

/**
 * Finds if this is a RemirrorProvider (which provides the RemirrorInjectedProps into the context);
 *
 * @param value - the value to check
 */
export const isRemirrorProvider = isRemirrorElementOfType(RemirrorElementType.EditorProvider);

/**
 * Checks if this is a ManagedRemirrorProvider which pulls in the manager from the context and places it's children
 * inside the RemirrorProvider
 *
 * @param value - the value to check
 */
export const isManagedRemirrorProvider = isRemirrorElementOfType(RemirrorElementType.ManagedEditorProvider);

/**
 * Clones an element while also enabling the css prop on jsx elements at the same time.
 * This is used for emotion which needs to inject the css property which React.cloneElement doesn't support.
 *
 * @param element - the element to clone
 * @param props - the props to pass through to the cloned element
 * @param rest - the children of the cloned element
 *
 * @returns a cloned react element with builtin support for the emotion `css` props
 */
export const cloneElement = <GProps extends PropsWithChildren<{ ref?: LegacyRef<any> }> = any>(
  element: ReactElement<GProps>,
  props: GProps,
  ...rest: ReactNode[]
) => {
  const children = uniqueArray([
    ...(isArray(props.children) ? props.children : props.children ? [props.children] : []),
    ...(isArray(rest) ? rest : rest ? [rest] : []),
  ]);

  return jsx(
    element.type,
    {
      key: element.key,
      ref: element.props.ref,
      ...element.props,
      ...props,
    },
    ...children,
  );
};

/**
 * Will throw an error if the child provided is not a function.
 *
 * @remarks
 * This is currently used in the remirror component to throw an error when the element children
 * are not a render prop. It should be called outside of render for class Components.
 *
 * @param prop - the prop to test
 */
export const propIsFunction = (prop: unknown): prop is AnyFunction => {
  if (!isFunction(prop)) {
    throw new Error('The child argument to the Remirror component must be a function.');
  }
  return true;
};
