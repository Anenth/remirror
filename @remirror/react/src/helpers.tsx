import { ExtensionManager, PlainObject } from '@remirror/core';
import React, { ComponentType } from 'react';
import { renderToString } from 'react-dom/server';
import { useRemirrorManager } from './hooks';

export interface GetManagerFromComponentTreeParams {
  /**
   * The full remirror component tree wrapped with an outer <RemirrorManager />
   */
  Component: ComponentType<any>;

  /**
   * A prop for inserting the component which will retrieve the manager
   * @default 'children'
   */
  prop?: string;

  /**
   * Other props that need to be passed into the component
   */
  extraProps?: PlainObject;
}

/**
 * Retrieves the extension manager from a component tree.
 *
 * @remarks
 * When building your multiplatform editor you will want to only use once schema across
 * all your environments. Since Prosemirror only renders in the DOM it makes sense to set
 * up the schema in your DOM based react code.
 *
 * But how to use that same schema across your platforms?
 *
 * This function provides a way of doing just that. It just requires that you set up your
 * editor with with a wrapper component that uses the `RemirrorManager` and then provides
 * a prop (`children` by default) to place a component within the `RemirrorManager`.
 *
 * This function uses the `prop` you provide to insert a component whose sole purpose is
 * to pull the extension manager from the React context and pass it back to the consumer
 * of this function.
 *
 * ```ts
 * // client.ts
 * export const MyEditorWrapper = ({ children }) => {
 *   return (
 *     <RemirrorManager>
 *       {children}
 *       <RemirrorExtension Constructor={NewExtension} />
 *       <RemirrorExtension Constructor={PlaceholderExtension} emptyNodeClass='empty' />
 *     </RemirrorManager>
 *   );
 * };
 *
 * // server.ts
 * import { MyEditorWrapper } from './client';
 *
 * const manager = await getManagerFromComponentTree({
 *   Component: <MyEditorWrapper />,
 *   prop: 'children',
 * });
 *
 * // For example now you have access to the schema
 * const { schema } =  manager;
 *
 * ```
 *
 * This is useful in server side rendered environments.
 */
export const getManagerFromComponentTree = ({
  Component,
  prop = 'children',
  extraProps = {},
}: GetManagerFromComponentTreeParams) =>
  new Promise<ExtensionManager>((resolve, reject) => {
    const MangerRetriever = () => {
      const manager = useRemirrorManager();
      resolve(manager);
      return <></>;
    };
    const props = { ...extraProps, [prop]: <MangerRetriever /> };
    renderToString(<Component {...props} />);
    reject(
      new Error(
        `The manager was not found. Please check that \`${Component.displayName ||
          Component.name}\` has a prop called \`${prop}\` which is rendered within the \`<RemirrorManager />\` context`,
      ),
    );
  });
