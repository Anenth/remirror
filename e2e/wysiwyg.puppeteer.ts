import { Deployment, URLDescriptor } from '@test-fixtures/test-urls';
import { outerHtml } from './helpers';

const editorSelector = '.remirror-editor';

describe('SSR Twitter Showcase', () => {
  beforeEach(async () => {
    await jestPuppeteer.resetPage();
    // Set JavaScript to disabled to mimic server side rendering
    await page.setJavaScriptEnabled(false);
    await page.goto(URLDescriptor.wysiwyg[Deployment.Next][1]);
  });

  it('should pre render the editor', async () => {
    // This test checks that the ProsemirrorEditor exists and contains the expected html
    await expect(outerHtml(editorSelector)).resolves.toBeTruthy();
    await expect(outerHtml(editorSelector)).resolves.toMatchInlineSnapshot(`
            "<div role=\\"textbox\\" aria-multiline=\\"true\\" aria-label=\\"\\" class=\\"remirror-editor\\"><h1>A WYSIWYG Editor</h1><p></p><p>With <strong><em>formatting </em></strong>and other <u>enhancements</u>.</p><p></p><pre>Simple Code Blocks
            echo \\"fun times\\"
            Use Shift-Enter or Mod-Enter to hard break out of the code block<code></code></pre><p></p><ul><li><p>List support</p><ul><li><p>With nesting</p></li></ul></li><li><p>Built in.</p></li></ul><p></p><blockquote><p>Block quotes</p><h3>With support for headers</h3></blockquote><p></p><p>And <a href=\\"https://google.com\\" rel=\\"noopener noreferrer nofollow\\">urls</a> with support for <code>Cmd-k</code> shortcut to enter a link manually.</p><p></p><p></p></div>"
          `);
  });
});
