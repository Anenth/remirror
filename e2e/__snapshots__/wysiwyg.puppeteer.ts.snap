// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wysiwyg Editor Snapshots SSR should pre render the editor 1`] = `
"<div role=\\"textbox\\" aria-multiline=\\"true\\" aria-label=\\"\\" aria-placeholder=\\"Start typing...\\" class=\\"Prosemirror remirror-editor\\" contenteditable=\\"true\\"><h1>A WYSIWYG Editor</h1><p><br></p><p>With <strong><em>formatting </em></strong>and other <u>enhancements</u>.</p><p><br></p><pre class=\\"language-markdown\\"><code><span class=\\"token title important\\"><span class=\\"token punctuation\\">##</span> Simple Code Blocks</span>

<span class=\\"token code\\"><span class=\\"token punctuation\\">\`\`\`</span><span class=\\"token code-language\\">js</span>
<span class=\\"token code-block language-js\\">console<span class=\\"token punctuation\\">.</span><span class=\\"token function\\">log</span><span class=\\"token punctuation\\">(</span><span class=\\"token string\\">\\"with code fence support\\"</span><span class=\\"token punctuation\\">)</span><span class=\\"token punctuation\\">;</span></span>
<span class=\\"token punctuation\\">\`\`\`</span></span>

<span class=\\"token code\\"><span class=\\"token punctuation\\">\`\`\`</span><span class=\\"token code-language\\">bash</span>
<span class=\\"token code-block language-bash\\"><span class=\\"token keyword\\">echo</span> <span class=\\"token string\\">\\"fun times\\"</span></span>
<span class=\\"token punctuation\\">\`\`\`</span></span>

Use Shift-Enter or Mod-Enter to hard break out of the code block</code></pre><p><br></p><ul><li><p>List support</p><ul><li><p>With nesting</p></li></ul></li><li><p>Built in.</p></li></ul><p><br></p><blockquote><p>Block quotes</p><h3>With support for headers</h3></blockquote><p><br></p><p>And <a href=\\"https://google.com\\" rel=\\"noopener noreferrer nofollow\\">urls</a> with support for <code>Cmd-k</code> shortcut to enter a link manually.</p><p><br></p><p><br></p></div>"
`;
